syntax = "proto3";

package shopping;

service Market {
    rpc RegisterSeller(SellerInfo) returns (Response);
    rpc SellItem(ItemInfo) returns (Response);
    rpc UpdateItem(ItemInfo) returns (Response);
    rpc DeleteItem(ItemID) returns (Response);
    rpc DisplaySellerItems(SellerInfo) returns (ItemList);
    rpc SearchItem(SearchRequest) returns (ItemList);
    rpc BuyItem(BuyRequest) returns (Response);
    rpc AddToWishList(WishListRequest) returns (Response);
    rpc RateItem(RatingRequest) returns (Response);
    rpc NotifyClient(NotificationRequest) returns (Response);
}

enum Category {
    ELECTRONICS = 0;
    FASHION = 1;
    BOOK = 2;
    OTHERS = 3;
    ANY = 4;
}

message SellerInfo {
    string address = 1;
    string uuid = 2;
}

message ItemInfo {
    int32 id = 1;
    string name = 2;
    Category category = 3;
    int32 quantity = 4;
    string description = 5;
    string seller_address = 6;
    float price = 7;
    string seller_uuid = 8;
}

message ItemID {
    int32 id = 1;
    string seller_address = 2;
}

message Response {
    string message = 1;
}

message Item {
    int32 id = 1;
    float price = 2;
    string name = 3;
    Category category = 4;
    string description = 5;
    int32 quantity = 6;
    string seller_address = 7;
    float rating = 8;
}

message SearchRequest {
    string name = 1;
    Category category = 2;
}

message ItemList {
    repeated Item items = 1;
}

message BuyRequest {
    int32 item_id = 1;
    int32 quantity = 2;
    string buyer_address = 3;
}

message WishListRequest {
    int32 item_id = 1;
    string buyer_address = 2;
}

message RatingRequest {
    int32 item_id = 1;
    int32 rating = 2;
    string buyer_address = 3;
}

message NotificationRequest {
    int32 item_id = 1;
    string buyer_address = 2;
}
